File: DataStructure/linked_list/linked_list.java
package practice;

class Node{
	public int data;
	public Node next;
	
	public Node() {
		this.data = 0;
		this.next = null;
	}
	public Node(int data) {
		this.data = data;
		this.next = null;
	}
	public Node(int data, Node node) {
		this.data = data;
		this.next = node;
	}
}

public class linked_list {
	private Node head;
	public int nodeCount = 0;
	 
	public linked_list() {
		head = null;
	}
	
	// add data
	public void Append(int data) {
		Node newNode = new Node(data);
		if(head == null) {
			this.head = newNode;
		}
		else {
			Node tmp = head;
			while(tmp.next != null) {
				tmp = tmp.next;
			}
			tmp.next = newNode;
		}
		nodeCount++;
	}
	
	public void Insert(int pos, int data) {
		if(pos > nodeCount) {
			System.out.println("error\n");
			System.exit(0);
		}
		Node newNode = new Node(data);
		
		Node prev_node = getAt(pos);
		if(prev_node == null) {
			prev_node = getAt(pos-1);
			prev_node.next = newNode;
		}
		else {
			newNode.next = prev_node.next;
			prev_node.next = newNode;
		}
		nodeCount++;
	}
	
	public Node getAt(int pos) {
		int cnt = 0;
		Node tmp = head;
		Node return_node = null;
		
		while(cnt < nodeCount) {
			if(cnt == pos)
				break;
			tmp = tmp.next;
			cnt++;
		}
		if(cnt < pos)
			return_node = null;
		else
			return_node = tmp;
		return return_node;
	}
	
	// delete data
	public void Delete(int pos) {
		Node delete_node = getAt(pos);
		if(delete_node == null) {
			System.out.println("error\n");
			System.exit(0);
		}
		
		if(pos == 0) {
			head = head.next;
			delete_node = null;
		}
		else {
			Node prev_node = getAt(pos - 1);
			if(delete_node.next == null) {
				prev_node.next = null;
				delete_node = null;
			}
			else {
			Node next_node = getAt(pos + 1);
			prev_node.next = next_node;
			delete_node = null;
			}
		}
		nodeCount--;
	}
	
	public void printlist() {
		Node tmp = head;
		do{
			System.out.print(tmp.data + "->");
			tmp = tmp.next;
		} while(tmp.next != null); 
		System.out.println(tmp.data);
	}
	
	
	public static void main(String args[]) {
		linked_list List = new linked_list();
		
		List.Append(5);
		List.Append(6);
		List.Append(8);
		List.Append(12);
		List.printlist(); // 5->6->8->12
		
		List.Insert(2, 29);
		List.printlist(); // 5->6->8->29->12
		List.Insert(5, 39);
		List.printlist(); // 5->6->8->29->12->39
		List.Insert(4, 9);
		List.printlist(); // 5->6->8->29->12->9->39
		
		List.Delete(5);
		List.printlist(); // 5->6->8->29->12->39
		List.Delete(5);
		List.printlist(); // 5->6->8->29->12
		List.Delete(0);
		List.printlist(); // 6->8->29->12
	}
}

File: DataStructure/linked_list/linked_list.py
class node:
	def __init__(self, item):
		self.data = item
		self.nextNode = None
	
class LinkedList:
	def __init__(self):
		self.head = None
		self.tail = None
		self.nodeCount = 0
	
	def Append(self, item):
		newNode = node(item)
		if self.nodeCount == 0:
			self.head = newNode
			self.tail = newNode
		else:
			self.tail.nextNode = newNode
			self.tail = newNode
		self.nodeCount += 1
		
	def getAt(self, pos):
		cnt = 0
		Node = self.head
		while Node is not None:
			current = Node
			if pos == cnt:
				return current
			Node = Node.nextNode
			cnt += 1
	
	
	def Delete(self, pos):
		if pos < self.nodeCount:
			Delete_Node = self.getAt(pos)
		else:
			Delete_Node = self.tail
			
		if pos == 0:
			self.head = Delete_Node.nextNode
			Delete_Node.nextNode = None
		else:
			prev_Node = self.getAt(pos - 1)
			prev_Node.nextNode = Delete_Node.nextNode
			Delete_Node.nextNode = None
			if pos == self.nodeCount:
				self.tail = prev_Node
			
		print("Delete(%d, %d)" %(pos, Delete_Node.data))
		self.nodeCount -= 1
		
	def Insert(self, pos, item):
		Insert_Node = node(item)
		if pos == 0:
			current_Node = self.head
			Insert_Node.nextNode = current_Node
			self.head = Insert_Node
		else:
			prev_Node = self.getAt(pos - 1)
			current_Node = prev_Node.nextNode
			
			prev_Node.nextNode = Insert_Node
			Insert_Node.nextNode = current_Node
		
		if pos > self.nodeCount:
			tmp_Node = self.tail
			tmp_Node.nextNode = Insert_Node
			self.tail = Insert_Node
			
		print("Insert(%d, %d)" %(pos, item))
		self.nodeCount += 1
		
	def PrintList(self):
		Node = self.head
		for idx in range(self.nodeCount):
			if Node.nextNode is not None:
				print(Node.data, end = " -> ")
			else:
				print(Node.data)
			Node = Node.nextNode
	
	

if __name__ == "__main__":
	L = list(map(int, input().strip().split(' ')))
	Linked = LinkedList()
	for data in L:
		Linked.Append(data)
	Linked.PrintList()
	
	# Delete Node
	Linked.Delete(0)	
	Linked.PrintList() # 10 -> 15 -> 20 -> 25 -> 30
	Linked.Delete(4)
	Linked.PrintList() # 10 -> 15 -> 20 -> 25
	Linked.Delete(2)
	Linked.PrintList() # 10 -> 15 -> 25
	
	# Insert Node
	Linked.Insert(0, 6)
	Linked.PrintList() # 6 -> 10 -> 15 -> 25
	Linked.Insert(4, 30)
	Linked.PrintList() # 6 -> 10 -> 15 -> 25 -> 30
	Linked.Insert(3, 27)
	Linked.PrintList() # 6 -> 10 -> 15 -> 27 -> 25 -> 30
